crates/core/src/error.rs

```diff
@@ -1,810 +1,451 @@
+//!
+//! ### Conversion and Nested Results
+//!
+//! Converting between different error kinds or their variants e.g. `TreeErrorKind::Channel` -> `ArchiverErrorKind::Channel`
+//! should seldom happen (probably never?), as the caller is most likely not setup to handle such errors from a different layer,
+//! so at this point, we should return either a `RusticError` indicating this is a hard error. Or use a nested Result, e.g.
+//! `Result<Result<T, TreeErrorKind>, RusticError>`.
+//!
+//! Local error types in `pub fn` (associated) functions need to be manually converted into a `RusticError` with a good error message
+//! and other important information, e.g. actionable guidance for the user.
+//!
+//! ### Backend traits
+//!
+//! By using `RusticResult` in our `Backend` traits, we also make sure, we get back presentable errors for our users.
+//! We had them before as type erased errors, that we just bubbled up. Now we can provide more context and guidance.
+//!
+//! ### Traits
+//!
+//! All traits and implementations of (foreign) traits should use `RusticResult` as return type or `Box<RusticError>` as `Self::Err`.
+//!
+//! ### Display and Debug
+//!
+//! All types that we want to attach to an error should implement `Display` and `Debug` to provide a good error message and a nice way
+//! to display the error.
 
 // FIXME: Remove when 'displaydoc' has fixed/recommended further treatment upstream: https://github.com/yaahc/displaydoc/issues/48
 #![allow(clippy::doc_markdown)]
 // use std::error::Error as StdError;
 // use std::fmt;
 
+use ::std::convert::Into;
+use smol_str::SmolStr;
 use std::{
-    error::Error,
-    ffi::OsString,
-    num::{ParseFloatError, ParseIntError, TryFromIntError},
-    ops::RangeInclusive,
-    path::{PathBuf, StripPrefixError},
-    process::ExitStatus,
-    str::Utf8Error,
+    backtrace::Backtrace,
+    borrow::Cow,
+    fmt::{self, Display},
 };
 
-#[cfg(not(windows))]
-use nix::errno::Errno;
+pub(crate) mod constants {
+    pub const DEFAULT_DOCS_URL: &str = "https://rustic.cli.rs/docs/errors/";
+    pub const DEFAULT_ISSUE_URL: &str = "https://github.com/rustic-rs/rustic_core/issues/new";
+}
 
-use aes256ctr_poly1305aes::aead;
-use chrono::OutOfRangeError;
-use displaydoc::Display;
-use thiserror::Error;
+/// Result type that is being returned from methods that can fail and thus have [`RusticError`]s.
+pub type RusticResult<T, E = Box<RusticError>> = Result<T, E>;
 
-use crate::{
-    backend::node::NodeType,
-    blob::{tree::TreeId, BlobId},
-    repofile::{indexfile::IndexPack, packfile::PackId},
-};
+/// Severity of an error, ranging from informational to fatal.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Severity {
+    /// Informational
+    Info,
 
-/// Result type that is being returned from methods that can fail and thus have [`RusticError`]s.
-pub type RusticResult<T> = Result<T, RusticError>;
+    /// Warning
```

Comment:
two questions about `Severity` : 
- is `Info` really a possible severity for our errors?
- `Severity` seems to never be used in this refactor (apart from re-exporting), should we keep it?

About both these questions I feel like `RusticError` is overall a non-recoverable error:
- we have nested `Result`s for recoverable errors
- `RusticError` is not made to be programmatically dealt with, it is an error to the user. You cannot `match` a rustic error more than the rather large `ErrorKind` enum.

For this reason I feel like a `RusticError` is always `Error` if not `Fatal`.

Reply:
Yeah, severity might be useless overall. As well as status maybe. I was thinking of having it, so we can use it down the line without breaking our API for these things. Though that shouldn't be the case, now that we use `non_exhaustive`. So for sure, we can remove that.